src/layers/basic.jl

```diff
@@ -503,3 +503,56 @@ function Base.show(io::IO, e::Embedding)
 end
 
 outputsize(e::Embedding) = (e.out_dims,)
+
+"""
+    Periodic(dims, periods)
+
+Create a layer periodic in `dims` with respective periods `periods`. All other input
+dimensions are passed through unchanged, but `dims` are replaced with their sine and cosine
+(scaled appropriately to have the specified periods).
+
+## Arguments
+
+  - `dims`: Periodic dimensions
+  - `periods`: Periods of the dimensions `dims`, in the same order
+
+## Inputs
+
+  - `x` must be an AbstractArray with `issubset(dims, axes(x, 1))`
+
+## Returns
+
+  - AbstractArray with dimensions `(size(x, 1) + length(dims), ...)` where `...` are the
+    other dimensions of `x`; the unchanged dimensions are first, then all the sines, then
+    all the cosines.
+  - Empty `NamedTuple()`
+"""
+@concrete struct Periodic <:AbstractExplicitLayer
+    dims
+    periods
+end
+
+
+@inline function (p::Periodic)(x::AbstractVector, ps, st::NamedTuple)
+    return vec(first(p(reshape(x, :, 1), ps, st))), st
+end
+
+@inline function (p::Periodic)(x::A, ps, st::NamedTuple) where A <:AbstractMatrix
+    k = convert(A, reshape(2 ./ p.periods, :, 1))
```

Comment:
I'm seeing an error from the `@jet` test that I think is coming from this line, but I haven't used that test before and am not experienced debugging those issues. Do you see the issue? Did I get the right problem line?

Reply:
make `dims` into a tuple and hoist the `periods` into the states. That way it is guaranteed to have the same device and such as `x`. You definitely don't want to do a `convert(A, ...)`

Reply:
Why make `dims` a tuple? The call to `view` doesn't like that. 

There's also some ambiguity in the word "dimension" here, since it can both refer to each row in `x` (e.g., the number of input dimensions of the layer) and also to `N` in `AbstractArray{T, N}` (an `AbstractMatrix` is two-dimensional regardless of how many rows and columns it has, which are the matrix's dimensions). Is there a clear way to distinguish between them? `dims` is referring to the first one. Should I replace it with "indices?"

Reply:
So, even without the call to `convert`, I'm still getting an error from that `@jet` and `@test_gradients` tests that I don't know how to diagnose or resolve. Any tips? Do you see what's going wrong?

Reply:
Run `@report_call` and `report_opt` from JET, it will tell you where the problem is coming from.

For `@test_gradients` can you show the failure? It should print which backend is not matching against Zygote

Reply:
Looking through the `@reportcall` output, and doing a bit of debugging, it looks like (perhaps among other issues) `view(x, any_idxs, :)` or similarly `x[any_idxs, :]` causes some kind of type-inference problem on the GPU according to JET. I'm not really sure where to go from there, since we need to be able to select only some rows of the input since we want to be able to create networks that are periodic in only some of the inputs. 

It's possible that the type issues are actually what's causing the error when it tries running the `@test_gradients` test, since the error is a MethodError about an ambiguous function call (two different versions of `_unsafe_setindex!`) when Tracker is calculating the gradient. 

Is there some other way I'm supposed to grab some rows, but not others? The layer is supposed to do something like `[x, y, z] -> [x, z, sin(y), cos(y)]`, so not being able to select a subset of the rows is basically a non-starter.

I appreciate the help, as I'm not very well-versed on the whole multiple-dispatch/type-stability stuff, especially when you mix in the GPU.

Reply:
I see, if the inference problem is purely downstream, i.e. if you set `target_modules` to be Lux,LuxLib,LuxCore then it shouldn't really be a concern. It does seem test_opt fails even with target modules set https://buildkite.com/julialang/lux-dot-jl/builds/1930#018f59fb-fdbf-4b03-8902-985fd77bca10/333-1009.

I don't like this solution but this would work:

```julia
@inline __maybe_view(x, args...) = view(x, args...)  # put in utils.jl
@inline Lux.__maybe_view(x::TrackedArray, args...) = x[args...]  # put in the TrackerExt file
```

then instead of view use `__maybe_view`, this won't be optimal for Tracker but that is fine

Reply:
It looks like removing calls to `view` solved all the JET and gradient problems. Any way I tried having a view caused some kind of type issue with CUDA. I know it's not ideal, but hopefully it won't impact performance too much.